### PR TITLE
Add Makefile from crictl proj

### PR DESCRIPTION
**Pull request by soharaki**
Add Makefile from crictl proj. Makefile runs building and linting
`nerdctl` automatically. #2

Signed-off-by: Sho Haraki <endurance.test.20380119@gmail.com>

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,2 @@
 /nerdctl
+build
```
Comment (AkihiroSuda) [inline .gitignore:2]: You meant _output?

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+GO ?= go
+
+# test for go module support
+ifeq ($(shell go help mod >/dev/null 2>&1 && echo true), true)
+export GO_BUILD=GO111MODULE=on $(GO) build -mod=vendor
```
Comment (AkihiroSuda) [inline Makefile:21]: This project does not use vendor

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+export GO_TEST=GO111MODULE=on $(GO) test -mod=vendor
+else
+export GO_BUILD=$(GO) build
+export GO_TEST=$(GO) test
+endif
+
+GOOS := $(shell $(GO) env GOOS)
+ifeq ($(GOOS),windows)
```
Comment (AkihiroSuda) [inline Makefile:29]: Windows is unlikely going to be supported (unless somebody contributes)

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+	BIN_EXT := .exe
+endif
+
+PROJECT := github.com/AkihiroSuda/nerdctl
+BINDIR ?= /usr/local/bin
+
+VERSION := $(shell git describe --tags --dirty --always)
+VERSION := $(VERSION:v%=%)
+GO_LDFLAGS := -X $(PROJECT)/pkg/version.Version=$(VERSION)
```
Comment (AkihiroSuda) [inline Makefile:38]: Nonexistent pkg

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+BUILDTAGS := selinux
```
Comment (AkihiroSuda) [inline Makefile:39]: Unused build tag

File: Makefile [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+BUILD_PATH := $(shell pwd)/build
```
Comment (AkihiroSuda) [inline Makefile:41]: Doesn't seem to be needed

File: Makefile [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+BUILD_BIN_PATH := $(BUILD_PATH)/bin
+
+define go-build
+	$(shell cd `pwd` && $(GO_BUILD) -o $(BUILD_BIN_PATH)/$(shell basename $(1)) $(1))
+	@echo > /dev/null
+endef
+
+GINKGO := $(BUILD_BIN_PATH)/ginkgo
```
Comment (AkihiroSuda) [inline Makefile:49]: This project does not use ginkgo

File: Makefile [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+	curl -sfL $$URL/$$VERSION/install.sh | sh -s $$VERSION
+
+
+$(GINKGO):
+	$(call go-build,./vendor/github.com/onsi/ginkgo/ginkgo)
+
+vendor:
+	export GO111MODULE=on GOSUMDB= && \
+		$(GO) mod tidy && \
+		$(GO) mod vendor && \
```
Comment (AkihiroSuda) [inline Makefile:101]: This project does not use vendor

File: hack/verify-go-modules.sh [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -e
+
+export GO111MODULE=on
+go mod tidy && go mod vendor && go mod verify
```
Comment (AkihiroSuda) [inline hack/verify-go-modules.sh:22]: No vendor

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]
Thanks, but looks too complex

File: .gitignore (hunk 1 above)
Comment (soharaki) [reply .gitignore:2]: it's for the golangcli-lint. It's just downloaded there.

File: Makefile (hunk 2 above)
Comment (soharaki) [reply Makefile:21]: Thank you.
Well, not use vendor? (I actually "make build" without vendor, and I got the error "containerd not found etc.."
Would you tell me how to build this with some option?

File: Makefile (hunk 3 above)
Comment (soharaki) [reply Makefile:29]: OK. I will fix it.

File: Makefile (hunk 4 above)
Comment (soharaki) [reply Makefile:38]: Thank you, I will delete it.

File: Makefile (hunk 5 above)
Comment (soharaki) [reply Makefile:39]: Thank you, I will delete it.

File: Makefile (hunk 7 above)
Comment (soharaki) [reply Makefile:49]: Thank you, I will delete it.

**Comment by soharaki** [comment]
Thank you for the review.
I think it writing more simply too.

File: Makefile (hunk 2 above)
Comment (AkihiroSuda) [reply Makefile:21]: `go build` should work with Go 1.15

**Comment by soharaki** [comment]
@AkihiroSuda 
Thank you for the review. I fixed the all point.
Would you re-review the Makefile and .gitignore?

File: Makefile [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+.PHONY: \
+	help \
+	nerdctl \
+	clean \
+	binaries \
+	install \
+	install-nerdctl \
+	lint \
+	install.tools
```
Comment (inductor) [inline Makefile:84]: Add new line at EOF?

File: hack/verify-go-modules.sh [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+STATUS=$(git status --porcelain)
+if [[ -z $STATUS ]]; then
+    echo "tree is clean"
+else
+    echo "tree is dirty, please commit all changes"
+    echo ""
+    echo "$STATUS"
+    exit 1
+fi
```
Comment (inductor) [inline hack/verify-go-modules.sh:29]: ditto

**Comment by AkihiroSuda** [comment]
Thanks, but hack/verify-go-modules.sh seems no longer useful. Can we remove this file? Also, please squash commits.

**Comment by soharaki** [comment]
> Thanks, but hack/verify-go-modules.sh seems no longer useful. Can we remove this file? Also, please squash commits.

I squached commits and remove hack/verify-go-modules.sh and also add Makefile EOF by @inductor.

**Review by AkihiroSuda — APPROVED** [review]

**Pull request merged**